streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -363,7 +366,7 @@ private void maybeOptimizeRepartitionOperations() {
     private void maybeUpdateKeyChangingRepartitionNodeMap() {
         final Map<StreamsGraphNode, Set<StreamsGraphNode>> mergeNodesToKeyChangers = new HashMap<>();
         for (final StreamsGraphNode mergeNode : mergeNodes) {
-            mergeNodesToKeyChangers.put(mergeNode, new HashSet<>());
+            mergeNodesToKeyChangers.put(mergeNode, new LinkedHashSet<>());
```

Comment:
Keep the key-changing parent nodes for the merge node in order.

Reply:
ditto